### PR TITLE
SQL: Fix cyclic resolution of references to aliases in filters

### LINKED ISSUE #81577: [SQL] StackOverflowError from a query using aggregates and aliases
For an index with a single field in its mapping
```
    "sequence": {
      "type": "long"
    }
```
The following query results in a `StackOverflowError`:

```
select max(sequence) as sequence, avg(sequence) as sequence, avg(sequence) as s from test_index where s > 0 group by sequence having s > 0
```

The stacktrace:

```
java.lang.StackOverflowError: null                                                          
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:176) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
        at org.elasticsearch.xpack.ql.tree.Node.lambda$transformDown$8(Node.java:179) ~[?:?]
        at org.elasticsearch.xpack.ql.tree.Node.transformChildren(Node.java:211) ~[?:?]     
        at org.elasticsearch.xpack.ql.tree.Node.transformDown(Node.java:179) ~[?:?]         
...
```

### PR DESCRIPTION
**Pull request by Luegg**
Resolves #81577 and #81770

Both issues have been caused by `ResolveFilterRefs` eagerly resolving references replaced by aliases instead of delegating to the global name resolution.

E.g. `ResolveFilterRefs` used to resolve the condition in `SELECT gender emp_no, emp_no e FROM test_emp WHERE e > 10099` like this:

`e > 10099` => `emp_no > 10099` => `test_emp.gender > 10099`

Hence, it skipped the field resolution of `emp_no`.

Because an alias in a `SELECT` clause can reference a field of the same name (as in `MAX(emp_no) AS emp_no`), this behavior also caused infinite resolution loops as observed in #81577.

With this patch applied, the resolution will be:

`e > 10099` => `emp_no > 10099` => `test_emp.emp_no > 10099`

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/AnalyzerTests.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    public void testResolveAlternatingRecursiveFilterRefs() {
+        // Queries like the following used to cause a StackOverflowError in ResolveFilterRefs.
+        // see https://github.com/elastic/elasticsearch/issues/81577
+        LogicalPlan plan = analyze(
+            "SELECT MAX(emp_no) emp_no, MAX(emp_no) emp_no, AVG(emp_no) e FROM test WHERE e > 0 GROUP BY emp_no HAVING e > 0"
+        );
+
+        assertThat(plan, Matchers.instanceOf(Having.class));
+        Having h = (Having) plan;
+        assertThat(h.condition(), Matchers.instanceOf(GreaterThan.class));
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/AnalyzerTests.java:61]: Please, use static import for `Matchers` methods.

**Review by astefan — COMMENTED** [review]
Can you, please, add more tests?
I'd like to see flavors of SUB-SELECTS using this fix.

File: x-pack/plugin/sql/qa/server/src/main/resources/select.csv-spec [hunk 2]
```diff
@@ -241,4 +241,12 @@ Weiyi          |WEIYI
 Vishv          |VISHV          
 Valter         |VALTER         
 Valdiodio      |VALDIODIO
-;
\ No newline at end of file
+;
+
+ambiguousReferenceInWhereClause
+SELECT gender emp_no, emp_no e FROM test_emp WHERE e > 10099;
```
Comment (costin) [inline x-pack/plugin/sql/qa/server/src/main/resources/select.csv-spec:247]: You could add the raw column (gender) in the select to further reinforce the point:
`SELECT gender, gender emp_no, ...`
so the result contains both gender and emp_no with the same value.

**Review by costin — APPROVED** [review]
Nice fix. Surprised that we haven't ran into it before and we don't get a lot of failures.
+1 on more testing for subqueries with ambiguities.

**Comment by astefan** [comment]
@elasticmachine update branch

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by astefan** [comment]
@elasticmachine update branch

**Review by bpintea — APPROVED** [review]
LGTM

**Pull request merged**